### PR TITLE
add UGrid c-lib

### PR DESCRIPTION
**Pull request by ocefpaf**
Adding https://github.com/Deltares/UGrid

@rsignell-usgs this is the first step but UGridpy will be more challenging b/c that is not ready to use an external c-lib for UGrid yet. They are bundling a pre-built .so/.dll files it in their source code and that is a no-no is most packaging ecosystems.

TODO:
- [ ] patch UGridpy to use the c-lib
- [ ] package the `meshkernel` dependency

### CODE REVIEW

**Comment by ocefpaf** [comment]
@wolfv sorry to bug you but you are the only C++ expert I know. This one passes locally but it is failing in our CIs, any advice for me here? All of my patches here are "Google-SO" based and I have no idea of that I'm doing.

**Comment by wolfv** [comment]
indeed, weird error (can't easily reproduce the compiler error on Godbolt, either).

I have some ideas on what to try:

Easiest: just remove the unique_ptr in these tests, and use:

```
    const double[] node_y{0, 0, 1, 1, 2, 2, 3, 3, 0, 1, 2, 3, 0, 1, 2, 3};
    mesh2d.node_y = node_y;
...
```


Or, instead of `std::unique_ptr<double> const node_y(new double[]{...});` you could try `std::unique_ptr<double[]> const node_y(new double[]{...});` (not sure if that would make a difference ...).

File: recipes/UGrid/build.sh [hunk 1]
```diff
@@ -0,0 +1,11 @@
+#!/bin/sh
+
+mkdir build && pushd build
+cmake \
+    -D CMAKE_INSTALL_PREFIX=${PREFIX} \
```
Comment (marcelotrevisani) [inline recipes/UGrid/build.sh:5]:
```suggestion
    -D CMAKE_INSTALL_PREFIX=${PREFIX} \
    -D CXX_STANDARD=14 \
```

File: recipes/UGrid/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+requirements:
+  build:
+    - {{ compiler('cxx') }}
+    - cmake
+    - make
+  host:
+    - boost
+    - boost-cpp
+    - netcdf-cxx4
```
Comment (marcelotrevisani) [inline recipes/UGrid/meta.yaml:28]: you might need limb here as well
```suggestion
    - netcdf-cxx4
    - openlibm
```

Comment (marcelotrevisani) [inline recipes/UGrid/meta.yaml:28]:
```suggestion
    - netcdf-cxx4 >=4.3.1
```

File: recipes/UGrid/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  run:
+    - boost
+    - boost-cpp
+    - netcdf-cxx4
```
Comment (marcelotrevisani) [inline recipes/UGrid/meta.yaml:32]:
```suggestion
    - netcdf-cxx4 >=4.3.1
```

File: recipes/UGrid/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    - patches/use_runtime_error.patch
+
+build:
+  skip: True  # [win]
+  number: 0
+
+requirements:
+  build:
+    - {{ compiler('cxx') }}
+    - cmake
```
Comment (marcelotrevisani) [inline recipes/UGrid/meta.yaml:23]:
```suggestion
    - cmake >=3.19
```

**Comment by ocefpaf** [comment]
@rsignell-usgs both experts on cpp I know tried and failed here. I don't see the point on pursuing this one b/c upstream does not have an official release, proper CI, and transparent wheel builds. Also, both xugrid and uxarray, the former is also from Deltares btw, looks more promising and easier to use/package.

**Pull request closed without merging**